### PR TITLE
Week 3 Assignment - Advanced React and Databases

### PR DESCRIPTION
**Pull request by deenasun**
[//]: # "Feel free to customize this template to your project's vibes"

### Description
[//]: # "Required - Describe what's new in this PR in a few lines. A description and bullet points for specifics will suffice."

Created `components/AppContext.tsx`
- Context provider with functions to get all posts and get a post by its id
- `useEffect` to fetch all posts from supabase on first render
- `useCallback` and `useMemo` functions to memorize `getAllPosts` and `getPostById` functions based on `posts` state variable

Created `Feed` component
- Renders feed of posts from the home page
- Each post is clickable and redirects users to a page for that given post

Created `NavBar` component
- Contains the Blueprint logo and blog name
- Dynamic back button appears if a user is on a post-page and allows users to easily return to the home feed

Created `app/post/[id]/page.tsx` 
- Dynamic slug to render a specific post based on query parameter `id`
- Selects post from the AppContext provider
- Fetches and renders comments corresponding to the post underneath the main post component

Added `queryComentsByPostId` function in `supabase/client.ts`
- Fetches specific comments for a specific post based on the related `post_uuid`

Styling for the NavBar, back button, and Blueprint logo components

### Screenshots
[//]: # "Required for frontend changes, otherwise optional but strongly recommended. Add screenshots of expected behavior - GIFs if you're feeling fancy! Use the provided image template. Drag the desired image into the PR, then copy the link into the placeholder."

**Feed**

**Post page**

## How to review
[//]: # 'Required - Describe the order in which to review files and what to expect when testing locally. Is there anything specifically you want feedback on? Should this be reviewed commit by commit, or all at once? What are some user flows to test? What are some edge cases to look out for?'
```
git fetch origin
git checkout deenasun/week-3-assignment
pnpm run dev
```
Navigate to `http://localhost:3000`
You should see a scrollable feed of various posts.
Click on a post to navigate to a page just for that post and see its comments

## Next steps
[//]: # "Optional - What's NOT in this PR, doesn't work yet, and/or still needs to be done. Note any temporary fixes in this PR that should be cleaned up later."
Fonts got switched from Poppins for some reason

### Online sources
[//]: # 'Copy links to any tutorials or documentation that was useful to you when working on this PR'
- Next.js useRouter documentation: https://nextjs.org/docs/app/api-reference/functions/use-router

### Related PRs
[//]: # "Add related PRs you're waiting on/ PRs that will conflict, etc; if this is a refactor, feel free to add PRs that previously modified this code"

CC: @ronniebeggs

### CODE REVIEW

File: components/AppContext.tsx [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    return posts;
+  }, [posts]);
+
+  const getPostById = useCallback(
+    (id: string): QueryPost | null => {
+      const post = posts?.find(post => post.uuid === id);
+      return post ?? null;
+    },
+    [posts],
+  );
```
Comment (ronniebeggs) [inline components/AppContext.tsx:43]: Would it be better to store posts in an `id: QueryPost` map? It won't really matter when there are 3 posts but I think it'd be slightly more efficient.

File: components/Feed.tsx [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    return;
+  }
+
+  return (
+    <main className={styles.main}>
+      <NavBar showBack={false} />
+      {posts.map(posts => (
+        <div
+          key={posts.uuid}
+          onClick={() => router.push('/post/' + posts.uuid + '/')}
```
Comment (ronniebeggs) [inline components/Feed.tsx:34]: Very good!

File: components/NavBar.tsx [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+        <div className={styles.backButton} onClick={() => router.push('/')}>
+          <IoArrowBack />
+        </div>
+      )}
+      <div className={styles.bpLogo}>
+        <Image src={bpImage} width={30} height={30} alt="logo" />
+        blueprint blog
+      </div>
+    </div>
+  );
```
Comment (ronniebeggs) [inline components/NavBar.tsx:22]: This `NavBar` appears on both the Feed and detailed post viewing screens. Try using it within the `layout.tsx` file so it's applied to both screens with one call.

File: supabase/client.ts [hunk 4]
```diff
@@ -15,3 +15,12 @@ export async function queryComments() {
...
   return { comments, error };
 }
+
+export async function queryCommentsByPostId(postId: string) {
+  const { data: comments, error } = await supabase
+    .from('comments')
+    .select('*')
+    .eq('post_uuid', postId);
+
+  return { comments, error };
```
Comment (ronniebeggs) [inline supabase/client.ts:25]: Let's throw the `error` or at least `console.log` it if one appears. We don't have any error handling mechanisms in place, but it's good practice to throw errors in async queries.

File: components/AppContext.tsx [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+  createContext,
+  useCallback,
+  useEffect,
+  useMemo,
+  useState,
+} from 'react';
+import { queryPosts } from '@/supabase/client';
+import { QueryPost } from '@/types/types';
+
+interface AppContextState {
```
Comment (ronniebeggs) [inline components/AppContext.tsx:13]: I typically like to create a separate `/context` or `/utils` folder for declaring contexts.

File: components/Post.tsx [hunk 6]
```diff
@@ -32,19 +36,23 @@ export default function Post({
...
+      {imageUrl && (
+        <Image
+          alt="image of San Francisco"
+          className={styles.image}
+          width={300}
+          height={300}
+          src={imageUrl}
+        />
+      )}
       <InteractionBar numLikes={numLikes} />
```
Comment (ronniebeggs) [inline components/Post.tsx:56]: The Post component should have a different bottom tab bar depending on whether it's on the feed or post page. How can we get this to work?

File: components/Feed.tsx [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+import styles from '../app/styles.module.css';
+import '../assets/global.css';
+import { useContext, useEffect, useState } from 'react';
+import { useRouter } from 'next/navigation';
+import Post from '@/components/Post';
+import { QueryPost } from '@/types/types';
+import { AppContext } from './AppContext';
+import NavBar from './NavBar';
+
+export default function Feed() {
```
Comment (ronniebeggs) [inline components/Feed.tsx:12]: I don't think a feed component is necessary because it essentially defines the entire page. You might as well declare this logic in the `/app/page.tsx` file. Components should be smaller parts of the application that may be reusable.

**Review by ronniebeggs — CHANGES_REQUESTED** [review]
This looks very solid overall! I left a few minor comments but you're free to consider this complete once they're addressed!
